web/ui/react-app/src/pages/graph/Panel.tsx

```diff
@@ -138,55 +144,72 @@ class Panel extends Component<PanelProps, PanelState> {
         throw new Error('Invalid panel type "' + this.props.options.type + '"');
     }
 
-    fetch(`${this.props.pathPrefix}/${API_PATH}/${path}?${params}`, {
-      cache: 'no-store',
-      credentials: 'same-origin',
-      signal: abortController.signal,
-    })
-      .then(resp => resp.json())
-      .then(json => {
-        if (json.status !== 'success') {
-          throw new Error(json.error || 'invalid response JSON');
-        }
-
-        let resultSeries = 0;
-        if (json.data) {
-          const { resultType, result } = json.data;
-          if (resultType === 'scalar') {
-            resultSeries = 1;
-          } else if (result && result.length > 0) {
-            resultSeries = result.length;
-          }
+    let query;
+    let exemplars;
+    try {
+      query = await fetch(`${this.props.pathPrefix}/${API_PATH}/${path}?${params}`, {
+        cache: 'no-store',
+        credentials: 'same-origin',
+        signal: abortController.signal,
+      }).then(resp => resp.json());
+
+      if (query.status !== 'success') {
+        throw new Error(query.error || 'invalid response JSON');
+      }
+
+      if (showExemplars) {
+        exemplars = await fetch(`${this.props.pathPrefix}/${API_PATH}/query_exemplars?${params}`, {
+          cache: 'no-store',
+          credentials: 'same-origin',
+          signal: abortController.signal,
+        }).then(resp => resp.json());
+
+        if (exemplars.status !== 'success') {
+          throw new Error(exemplars.error || 'invalid response JSON');
         }
-
-        this.setState({
-          error: null,
-          data: json.data,
-          warnings: json.warnings,
-          lastQueryParams: {
-            startTime,
-            endTime,
-            resolution,
-          },
-          stats: {
-            loadTime: Date.now() - queryStart,
-            resolution,
-            resultSeries,
-          },
-          loading: false,
-        });
-        this.abortInFlightFetch = null;
-      })
-      .catch(error => {
-        if (error.name === 'AbortError') {
-          // Aborts are expected, don't show an error for them.
-          return;
+      }
+
+      let resultSeries = 0;
+      if (query.data) {
+        const { resultType, result } = query.data;
+        if (resultType === 'scalar') {
+          resultSeries = 1;
+        } else if (result && result.length > 0) {
+          resultSeries = result.length;
         }
-        this.setState({
-          error: 'Error executing query: ' + error.message,
-          loading: false,
-        });
+      }
+
+      if (exemplars && exemplars.data) {
+        query.data.exemplars = exemplars.data;
```

Comment:
I'd prefer using a completely separate `exemplars: Exemplar[]` state field for the exemplars, because right now `query` is just a dumb data transfer object that contains exactly the query response as delivered by the API. Adding extra stuff to it that wasn't returned in the same query could be confusing to readers.

Reply:
Agreed.